src/blueapi/startup/simmotor.py

```diff
@@ -90,5 +90,5 @@ def __init__(self, *, timeout: float, **kwargs) -> None:
         super().__init__(**kwargs)
         self._timeout = timeout
 
-    def set(self, value: float) -> Status:
-        return Status(timeout=self._timeout)
+    def set(self, value: float) -> MoveStatus:
```

Comment:
Should: No need to narrow the return type

```suggestion
    def set(self, value: float) -> Status:
```

Reply:
Pyright did complain, and it's a specific implementation. why no need?

Reply:
You shouldn't narrow the return type if the caller doesn't need it, you should keep the type general to preserve abstraction

Reply:
for whom? developer or the machine? IDE?

MoveStatus is the thing that the SimMotor returns. it gives me more information when I read a method to read that it's more narrow. 


you haven't provided a _source_ for your claim , but I can for mine - covariant return type

https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
https://en.wikipedia.org/wiki/Robustness_principle

Reply:
I do not believe that the robustness principle is the only principle to apply here, however this is quite a minor change and I don't hugely object to it going in. 

I think there is a question of policy for the codebase that we should address, probably outside this PR. I've just been down an interesting rabbit hole. I think this post sums up my point of view nicely: https://www.hacklewayne.com/postel-s-law-is-simple-to-fulfil-the-potential-of-your-code

In summary:
- The robustness principle (or Postel's law) is a tool that is used to maximise the usefulness of a given function to any potential caller
- Specifically you use it to vary the input and return types of the function, making the input as generic as possible and the output as specific as possible. The thing you do not vary is the implementation, since the input/return types are varied relative to that. The implementation performs a set of operations on the input type and provides a return type on which the caller can perform a different set of operations, therefore the variance is limited to input types that have the operations required for the implementation (but nothing else) and return types that have all the possible operations that the implementation can possibly provide.
- Usefulness is not the only metric to optimise for, however. We may also want to optimise for maintainability. By naively applying Postel's law you create functions that are as generic as possible in what they accept and as specific as possible in what they return relative to their implementation, this will make it harder to change the implementation in the future because callers may now depend on some of the large number of operations made available by the maximally specific return type. This is why it is better to have a return type with fewer possible operations if you are optimising for maintainability, you only provide the caller with what they really need. 

So this discussion is resolved by deciding how much we prioritise long-term maintainability, and therefore how much we do/don't apply Postel's law. In general I think maintainability is a very high priority for blueapi in its aim of avoiding the same problems that GDA has. I would rather effort were put in now to avoid pain later. As I say I think this discussion is out of scope for this PR so I'm happy to let this in, but for large-scale changes I will continue to advocate for the maintainability route. I'm also quite happy to continue this in an issue or other appropriate forum.

Reply:
> this will make it harder to change the implementation in the future because callers may now depend on some of the large number of operations made available by the maximally specific return type. This is why it is better to have a return type with fewer possible operations if you are optimising for maintainability, you only provide the caller with what they really need.


this is not a library though. this is an application. external callers are concerned with the REST api, not the internal code

Regarding this specific case this uses Ophyd so should be removed anyway. And a stronger argument is that ophyd-async in dodal and elsewhere just has generic `AsyncStatus`, not a specific `MoveStatus`.


> In general I think maintainability is a very high priority for blueapi in its aim of avoiding the same problems that GDA has. I would rather effort were put in now to avoid pain later

the goal of blueapi is to deliver business value in less than 5 years time. maintainability in 10 years time vs adaptability now is an opinionated tradeoff. Also this specific 'covariant return' aspect, both cases (for and against) can be construed as pursuing the *value* of maintainability, yet differ in the *strategy* chosen for it. 

_Trauma_ from past (software) experience can make (software) people overcorrect and act neurotically.


Specifically for GDA the three biggest pain points (afaik) - multiple repos to make one feature change, thick client, and chaotic display mechanisms - are all handled *for free* by the infrastructure and architecture (microservices and a separate web GUI). 

Not to mention improving organizational information flow practices. 

With all of this, marginal effectiveness of any more 'putting effort now' decreases.

> . By naively applying Postel's law you create functions that are as generic as possible in what they accept and as specific as possible in what they return relative to their implementation, this will make it harder to change the implementation in the future because callers may now depend on some of the large number of operations made available by the maximally specific return type.

that is ad-absurdum, not what I said.

finally, if the function does not say that a Movable returns a MoveStatus, it would be on a client to double check if the returned value is actually from a MoveStatus or just Status - the logic is just moved elsewhere.

Reply:
Yes, it is an opinionated tradeoff, which is why I wanted a discussion. For this specific case I think we agree re: ophyd-async and `AsyncStatus`, so can close this thread.

Reply:
agreed